### ISSUE TITLE
flatpak packaging

### ISSUE THREAD

**Issue by bam80** [sqlitebrowser/sqlitebrowser#2103]
Describe the new feature
--------------------------
[Snap package]( https://snapcraft.io/sqlitebrowser) already exists but not every distro is capable to run Snap, mostly because systemd is not accepted everywhere.
So flatpak package is very welcomed.

**Comment by scottfurry** [comment]
... or you could build from source.
Command line is your friend. ;)
`git clone [project git url]`
`cd sqlitebrowser`
`cmake -H. -Bbuild`
`make -C build`
`sudo make install`

**Comment by bam80** [comment]
The issue is not about I can't build. It's about adding smart distribution channel..

**Comment by scottfurry** [comment]
This sounds like a distribution problem. See availability [here]|(https://repology.org/project/sqlitebrowser/versions).

**Comment by bam80** [comment]
Flatpak as well as Snap are not tied to any distribution.

**Comment by tinywrkb** [comment]
[This](https://github.com/tinywrkb/flatpaks/blob/master/com.github.sqlitebrowser.sqlitebrowser/com.github.sqlitebrowser.sqlitebrowser.yaml) seems to work correctly.  
I once used the app and had it installed on my system. Instead of just removing the app from my system I converted it to a Flatpak.

Before anyone thinks about publishing this on Flathub please test that everything is working as expected.

**Comment by justinclift** [comment]
@scottfurry That "Repology" thing looks useful. :smile:

**Comment by scottfurry** [comment]
@justinclift - the one downside, after looking at the underlying code, is that it determines "latest" not from canonical source but who has "highest" revision. They fudge it otherwise. But it has great utility in looking at who has what revision.

**Comment by Qwertie-** [comment]
@tinywrkb I'll be testing out your flatpak config. If everything works fine, I think we should try to get it on flathub.

**Comment by tinywrkb** [comment]
@Qwertie- 

A few things to know about my manifest and Flatpak:

* The app seems to work fine in Wayland so it should use the `socket=fallback-x11` permission instead of `x11`. Only keep it if you really care about terribly outdated Flatpak versions. I updated the manifest accordingly. See more details [here](https://docs.flatpak.org/en/latest/flatpak-command-reference.html).
* The renaming directives are a workaround for the fact the app doesn't follow the reverse DNS schema as app naming convention. It would be better to fix this in the app build system.
* The `filesystem=home` permission is a little too permissive but maybe it's the best solution in order to avoid dealing with other issues, see next for details.
* Qt5 doesn't handle well the file portal when requesting access to folders. I hit this issue even when the app has file access permission to the folder / bind-mounted into the sandbox. [The documentation recommends to not use QFileDialog::DontUseNativeDialog](https://docs.flatpak.org/en/latest/portals-qt.html) but consider setting it as true when opening a file dialog to select a folder and see if it works correctly.
* Don't forget to read the [App Submission Guide](https://github.com/flathub/flathub/wiki/App-Submission).

**Comment by justinclift** [comment]
Oops, that was automatically closed.  Re-opening this, as it sounds like there's more to be done... :smile:

**Comment by Qwertie-** [comment]
@tinywrkb That's submitted to flathub now, works perfect on my machine. 

I couldn't get even single files to open correctly without home dir access so I guess for now that permission can stay.

I could not see any use for the permission `--device=dri`. Did you find something that needed it?

**Comment by bam80** [comment]
Could you also make sure it works on Wayland natively?

**Comment by tinywrkb** [comment]
@Qwertie-  

> That's submitted to flathub now, works perfect on my machine.

Great! 

> I couldn't get even single files to open correctly without home dir access so I guess for now that permission can stay.

There are XDG dirs permissions, like `xdg-documents` and etc, see `flatpak-overrride.1` man page.  
These filesystem permissions are basically bind mounts over folder/file targets in the sandbox, and only needed if the app requires to be able to directly access files without going through the File portal.  
The File portal is used when opening the file dialog and it returns a filehandle to a resource outside the sandbox, see [documentation](https://flatpak.github.io/xdg-desktop-portal/portal-docs.html#gdbus-org.freedesktop.portal.FileChooser).  
I really don't like setting `filesystem=home` permission, it too permissive as it also gives access to `.ssh`, `.gnupg` and etc, but sometimes as long as there's no Flathub policy in place this is up to the maintainer to decide.

> I could not see any use for the permission --device=dri. Did you find something that needed it?

`/dev/dri` is for 3D graphics, without it you fall back to software rendering, see [documentation](https://docs.flatpak.org/en/latest/sandbox-permissions.html).

**Comment by Qwertie-** [comment]
@tinywrkb 

>I really don't like setting filesystem=home permission, it too permissive as it also gives access to .ssh, .gnupg and etc,

Agreed but at the least, its no less secure than installing via dnf, etc. I think it should be ok for now until it works properly with portals. I'm not a big fan of the flatpaks which give specific access to Downloads/Documents etc. It just feels like bad UX when some of your files are inaccessible with no clear reason why.

>/dev/dri is for 3D graphics, without it you fall back to software rendering

Maybe I'm missing something here, but why would 3D graphics be needed here? Is the gui able to be gpu rendered if we have this permission?

**Comment by scottfurry** [comment]
>     I really don't like setting filesystem=home permission, it too
>     permissive as it also gives access to .ssh, .gnupg and etc,
>
> Agreed but at the least, its no less secure than installing via dnf, 
> etc. I think it should be ok for now until it works properly with 
> portals. I'm not a big fan of the flatpaks which give specific access 
> to Downloads/Documents etc. It just feels like bad UX when some of 
> your files are inaccessible with no clear reason why.
>
Anything in the user's home folder is fair game. If security is a 
primary concern, don't install to the home folder - use /opt or similar.
>
>     /dev/dri is for 3D graphics, without it you fall back to software
>     rendering
>
> Maybe I'm missing something here, but why would 3D graphics be needed 
> here? Is the gui able to be gpu rendered if we have this permission?
>
Hold up....
Why would a flatpak, which is suppose to be in a sandbox, need to access 
a *nix box's device folder (/dev)?
That seems wrong on multiple levels. If something needs to directly 
access device's 3D graphics in this manner, IMHO, there are bigger problems.

My $/€ 0.02

**Comment by tinywrkb** [comment]
> Maybe I'm missing something here, but why would 3D graphics be needed here? Is the gui able to be gpu rendered if we have this permission?

Why wouldn't it be needed? modern widget toolkits are using the GPU to render the GUI via 3D graphics APIs.  
All Flatpak GUI applications on Flathub have this permission enabled, this is the standard, just follow it. It's in the documentation I linked before.  
This is pretty low-level stuff, this is how the OpenGL drivers work, the only time I looked at code using it directly was when going through a Linux DRM guide (rendering not copyright). My guess is that you won't even be able to set the EGL context without it.

**Comment by Qwertie-** [comment]
@scottfurry I agree that home dir access is not the end of the world but its not the best that Flatpak offers. There is a feature of flatpak called "portals" where the GTK and QT filepickers are a semi privileged component that can see all of the files your user has access to, when the user clicks on a file in the file picker, the program inside flatpak is granted permission to access this file.

In theory this should mean you can use the program without any permissions to user files while to the user it looks like everything is normal. But I wasn't able to get that working. A future improvement perhaps.

Also I'm pretty sure that the `dri` permission does not give access to anything in the /dev dir other than /dev/dri which is apparently save and common to use.


@tinywrkb 
>All Flatpak GUI applications on Flathub have this permission enabled, this is the standard, just follow it. It's in the documentation I linked before.

Fair enough, I'll add that permission back in

**Comment by scottfurry** [comment]
> There is a feature of flatpak called "portals" where the GTK and QT 
> filepickers are a semi privileged component that can see all of the 
> files your user has access to, when the user clicks on a file in the 
> file picker, the program inside flatpak is granted permission to 
> access this file.
>
The curmudgeon in me is shaking my head...
All this noise and effort to get around installing a piece of software 
without installing needed dependencies.
New, hip, and/or wow != best
My grey hairs are hurting.

**Comment by frhun** [comment]
> All this noise and effort to get around installing a piece of software without installing needed dependencies. New, hip, and/or wow != best

I actually find that the best thing about flatpak by far is flathub, because it's a great software discoverability platform. And if you just want to try out a piece of software, without it cluttering your system, its quite perfect.
Your main every day use applications don't really have that much to gain from flatpak, other than graphical app store visibility. (and even that only on distributions where flathub is installed by default, or the user has manually installed it)
Also as a developer it can give you a rock solid foundation, and clear many common problems. For example when someone is running flatpak, you can be 100% sure about the environment your app is running in. (which is also why generally it is best if the application maintainers also maintain the flatpak, which is basically no effort at all - when the upstream build system is flexible enough; which is something other packagers will thank you for anyways)
You simply don't have to worry that some downstream package maintainer forgot to include some dependency that just happened to be installed on most systems, and consequently you won't get any tedious bug reports about these kind of issues.

**Comment by BeatLink** [comment]
I am not able to find this app on flathub, Was it not published?

**Comment by Qwertie-** [comment]
@BeatLink No it was not published. I had a go at it but never got all the way through review. I no longer have a linux system to test with so if someone else wants to pick it up, feel free.

**Comment by Mailaender** [comment]
I made an attempt to publish on https://github.com/flathub/flathub/pull/3110. You are invited to join.

**Comment by Mailaender** [comment]
https://github.com/flathub/org.sqlitebrowser.sqlitebrowser

**Comment by zer0-x** [comment]
As a verification system was added to flathub, I think it would be a great thing if the [flathub package](https://flathub.org/apps/org.sqlitebrowser.sqlitebrowser) of sqlitebrowser was verified.

For more: https://docs.flathub.org/docs/for-app-authors/verification/

**Comment by bam80** [comment]
> it would be a great thing if the [flathub package](https://flathub.org/apps/org.sqlitebrowser.sqlitebrowser) of sqlitebrowser was verified.

Ping:

<img width="1152" height="200" alt="Image" src="https://github.com/user-attachments/assets/1ea3892c-688b-4bea-950b-dcfbac152da1" />

**Comment by bam80** [comment]
@FriedrichFroebel why it's reopened?

**Comment by FriedrichFroebel** [comment]
@bam80 Because it was accidentally closed with a link to a comment in this issue and I do not know if all relevant aspects are resolved for this.

**Comment by bam80** [comment]
I think this particular issue is solved now. If something left, let's track it on separate issues.
Thank you to all contributors involved! Closing.
